content/blog/2020-12-19-LLVM IR to Assembly

```diff
@@ -0,0 +1,36 @@
+#Modifying LLVM’s IR to assembly on a existing ISA with different number of registers. 
+Most computer architectures have a fixed number of registers, and such number is often fixed and determined arbitrary.  Originally when 8086 was released (The original x86), there was only 8 registers and most of them are specialized, however in most recent iterations of x86-64 it has total of 40, including 16 general purpose registers, 8 mmx registers, and 16 SSE registers.  As the registers area overhead in modern out of order processors(O3) are relatively small, it is often common practice to incorporate more registers then what is absolutely minimum to complete the task.   Many advances topics in computing architecture such as simultaneous multithreading, Tomasulo algorithm, and speculative execution depend on the existence of having more physical registers then architectural registers.  
+This begs the question of do we actually need the number of registers offered by the modern ISA? Or can we go back to the traditional AX, BX, CX, DX, SI, DI, BP, SP registers?  Although at the current landscape in O-3 processors the incorporation of registers are relative cheap, such relationships may not hold true in other alternative processing methods.  Therefore such dimensions are worth a closer inspection, and that we as researcher should not take the cheap overhead for registers in modern O3 processors for granted. 
+This serves as the motivation of this project, in which I will modify LLVM assembly generation routine to generate code with less registers than what is offered by the traditional LLVM complier.  Furthermore, an informal comparison on the performance tradeoff based on the number of architectural registers is done.  
+#What did I do?
+Although initially the project wants to target the RISC-V ISA, but it is currently still experimental as of writing this blog post I have decided to use a more mature and widely supported language X86-64.  This project is largely done by modifying the LLVM x86 target configurations to create new target machine with different number of architectural registers.  Although X86-64 has better support across the LLVM stack, modifying the x86 machine target register configurations is no easy task and is significantly harder due to the number of aliasing it has across the different registers.   Most of the work that I have done on this project is modifying the LLVM assemble generation to a different number of registers.  Then by executing the complied output from gcc, llvm and modified llvm.  I hope to unearth the relationship of architectural registers and performance.  To compare the results across multiple register types, I have done microbenchmark using Matrix Multiplication with three different x86-64 implementation (scalar, SSE, AVX-128bit, AVX-256bit) and benchmarking with FFT computations.  
```

Comment:
The RISC-V extension is actually not experimental in LLVM anymore, as of LLVM 9. https://lists.llvm.org/pipermail/llvm-dev/2019-July/133724.html